Maple2.Server.Game/Manager/ShopManager.cs

```diff
@@ -85,162 +85,211 @@ private bool TryGetShopData(int shopId, [NotNullWhen(true)] out CharacterShopDat
+            Items = GetShopItems(metadata),
+        };
 
         if (!TryGetShopItemData(instancedShop.Id, out IDictionary<int, CharacterShopItemData>? dataDictionary)) {
-            dataDictionary = CreateShopItemData(serverShop);
+            dataDictionary = CreateShopItemData(instancedShop);
         }
 
         foreach ((int shopItemId, CharacterShopItemData itemData) in dataDictionary) {
-            if (!serverShop.Items.TryGetValue(shopItemId, out ShopItem? shopItem)) {
+            if (!instancedShop.Items.TryGetValue(shopItemId, out ShopItem? shopItem)) {
                 continue;
             }
 
-            ShopItem itemCopy = shopItem.Clone();
-            itemCopy.Item = itemData.Item;
-            itemCopy.StockPurchased = itemData.StockPurchased;
-            instancedShop.Items.Add(shopItemId, itemCopy);
+            instancedShop.Items[shopItemId] = new ShopItem(shopItem.Metadata) {
+                Item = itemData.Item,
+                StockPurchased = itemData.StockPurchased,
+            };
         }
 
         instancedShops[instancedShop.Id] = instancedShop;
         return instancedShop;
     }
 
-    private Shop CreateInstancedShop(Shop shop, CharacterShopData shopData, long restockTime) {
-        Shop instancedShop = shop.Clone()!;
-
+    private Shop CreateInstancedShop(ShopMetadata metadata, CharacterShopData shopData, long restockTime) {
+        var shop = new Shop(metadata);
         // First delete any existing shop item data
-        long ownerId = shop.RestockData!.PersistantInventory ? session.AccountId : session.CharacterId;
+        long ownerId = metadata.RestockData.AccountWide ? session.AccountId : session.CharacterId;
         using GameStorage.Request db = session.GameStorage.Context();
 
-        if (TryGetShopItemData(shop.Id, out IDictionary<int, CharacterShopItemData>? shopItemDatas)) {
+        if (TryGetShopItemData(metadata.Id, out IDictionary<int, CharacterShopItemData>? shopItemDatas)) {
             foreach ((int shopItemId, CharacterShopItemData itemData) in shopItemDatas) {
-                db.DeleteCharacterShopItemData(ownerId, shopItemId);
+                db.DeleteCharacterShopItemData(ownerId, itemData.ShopId, shopItemId);
             }
         }
 
         // Create new shop data
-        instancedShop.Items = GetShopItems(shop);
-        instancedShop.RestockTime = restockTime;
-        instancedShop.RestockData!.RestockCount = shopData.RestockCount;
-        CreateShopItemData(instancedShop);
-        instancedShops[instancedShop.Id] = instancedShop;
-        return instancedShop;
+        shop.Items = GetShopItems(metadata);
+        shop.RestockTime = restockTime;
+        shop.RestockCount = shopData.RestockCount;
+        CreateShopItemData(shop);
+        instancedShops[shop.Id] = shop;
+        return shop;
     }
 
-    private SortedDictionary<int, ShopItem> GetShopItems(Shop shop) {
+    private SortedDictionary<int, ShopItem> GetShopItems(ShopMetadata metadata) {
```

Comment:
**suggestion:** New GetShopItems method with probability-based item generation

The new GetShopItems method introduces probability-based item generation for shops with reset enabled. This is a significant change that adds complexity. Consider adding detailed comments explaining the logic, especially around the probability calculations and item count constraints.